Add sort listings functionality

taylor-yamashita opened the pull request:
Working with @ysl-ee to add sorting dropdown to "Items Requested" and "Items for Sale" gallery pages. (Taylor on frontend, Youngseo on backend)

Example demos: 

1. On the "Items Requested" page, user selects "price: low to high":
<img width="600" alt="Screen Shot 2023-03-21 at 12 13 56 PM" src="https://user-images.githubusercontent.com/80173363/226672175-bb407a6f-f523-4358-a4dc-e21752723f3f.png">

Result:
<img width="600" alt="Screen Shot 2023-03-21 at 12 14 12 PM" src="https://user-images.githubusercontent.com/80173363/226672889-fc1f3a54-5269-45dd-b90c-4b7f1bff84de.png">

2. On the "Items for Sale" page,  user selects "date posted: newest":
<img width="600" alt="Screen Shot 2023-03-21 at 12 15 29 PM" src="https://user-images.githubusercontent.com/80173363/226673618-5609e358-fe91-45e3-8cc6-13d2817b6a6d.png">

Result:
<img width="600" alt="Screen Shot 2023-03-21 at 12 15 40 PM" src="https://user-images.githubusercontent.com/80173363/226673669-53e20524-2833-44c6-a96e-691ea0f8a4d6.png">

ysl-ee commented:
- discovered that getItemsRelative() and getItemRequestsRelative() in views.py do not render directly do a page — they return a json reponse. we will probably have to return to these functions for the actual sorting functionality? we considered using a "sort_type" parameter in request.GET() so that we don't have to have 4 different URLs going on, but we removed it for the moment. (might bring back)
- gallery.html has updateQuery() which updates the query whenever new checkboxes (categories & conditions) are clicked. we are going to mimic this structure and implement the sorting right before it renders to prevent these selections from accidentally going away when we sort.
- category_pks and condition_indexes (from which the active ones are extracted) are declared at the top of the script 
- eventually, getItems() is kind of called (??) (i think and hope??) with a query string (see line 498 in populateItemsSynchronously())

taylor-yamashita commented:
^ Note: Generally, duplicate code is used to display available items vs. requested items, so we also put duplicate code in their respective places.

taylor-yamashita commented:
At the moment, everything is connected, but the sorting functions don't appear to behave as expected—e.g. sorting "price low to high" shows items with highest price first. @ysl-ee would you mind looking into this?

ysl-ee commented:
@taylor-yamashita sortings should be happening in correct order now :)

nicholaspad commented on TRT-django/marketplace/templates/marketplace/gallery.html:
```diff
@@ -63,11 +63,27 @@ <h5 style="font-size:30px" class="modal-title">Post Item</h5>
+              id="dropdownMenuButton" data-bs-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
+                sort by
+            </button>
+            <div class="dropdown-menu  dropdown-menu-right" aria-labelledby="dropdownMenuButton">
```
same nit about the spacing between class names

nicholaspad commented on TRT-django/marketplace/templates/marketplace/gallery.html:
```diff
@@ -177,6 +194,29 @@ <h5 style="font-size:30px" class="modal-title">Post Item</h5>
+    // sets sort_type based on current inputs, refresh listings
+    function updateQuerySort(sort_type_input) {
+      // update sort_type variable
+      if (sort_type_input == "price_hightolow") {
+          sort_type = "price_hightolow";
+      } else if (sort_type_input == "price_lowtohigh") {
+        sort_type = "price_lowtohigh";
+      } else if (sort_type_input == "date_oldtorec") {
+        sort_type = "date_oldtorec";
+      } else if (sort_type_input == "date_rectoold") {
+        sort_type = "date_rectoold";
+      }
```
same nit about setting `sort_type`

nicholaspad commented on TRT-django/marketplace/views.py:
```diff
@@ -465,16 +466,54 @@ def getItemsRelative(request):
+        sort_type = request.GET["sort_type"]
+    
+    if sort_type == "price_hightolow":
+        items = items.annotate(
+            row=Window(
+                expression=RowNumber(),
+                order_by=[F("price").asc(), F("pk").asc()], # also order by unique pk to make tie-breaks consistent
+            )
+        )
```
could you instead use this conditional block only to set the value passed into the `order_by` argument, and then make one call to `items.annotate` at the end?

```python
if sort_type == "price_hightolow":
    order_by = [F("price").asc(), F("pk").asc()]
elif ...:
    order_by = ...
elif ...:
    order_by = ...
elif ...:
    order_by = ...
else: ...
    order_by = ...

    # annotate items by search string rank
    items = items.annotate(rank=SearchRank(SearchVector("name", "description"), SearchQuery(search_string), cover_density=True))

items = items.annotate(
            row=Window(
                expression=RowNumber(),
                order_by=order_by, # also order by unique pk to make tie-breaks consistent
            )
        )
```

Danielpark1239 replied:
^ I'd also like to see the duplicate logic for items and item requests combined in some way

nicholaspad commented on TRT-django/marketplace/views.py:
```diff
@@ -1587,16 +1627,54 @@ def getItemRequestsRelative(request):
+    if "sort_type" in request.GET:
+        sort_type = request.GET["sort_type"]
+    
+    if sort_type == "price_hightolow":
+        item_requests = item_requests.annotate(
+            row=Window(
+                expression=RowNumber(),
+                order_by=[F("price").desc(), F("pk").asc()], # also order by unique pk to make tie-breaks consistent
+            )
+        )
```
same comment about shortening the conditional block

nicholaspad reviewed (changes requested):
awesome work @taylor-yamashita @ysl-ee, i would've found this one really difficult because the retail codebase is quite confusing

just a couple suggestions, but thanks for making this change!! a ton of people wanted this feature :)

shannon-heh commented on TRT-django/marketplace/templates/marketplace/browse_item_requests.html:
```diff
@@ -63,11 +63,27 @@ <h5 style="font-size:30px" class="modal-title">Post Item Request</h5>
+                <a class="dropdown-item" onClick="updateQuerySort('price_hightolow')">price: high to low</a>
+                <a class="dropdown-item" onClick="updateQuerySort('date_rectoold')">date posted: newest</a>
+                <a class="dropdown-item" onClick="updateQuerySort('date_oldtorec')">date posted: oldest</a>
+            </div>
```
good touch with keeping consistent with the lowercase :)

shannon-heh commented:
@ysl-ee @taylor-yamashita could one of you upload a short screen recording or a series of screenshots to demonstrate the working functionality? i'd also like to see what the drop-down looks like

shannon-heh reviewed (changes requested):
small nit: could you change cursor behavior when you hover a dropdown item? right now it's the default link cursor, but `cursor: pointer;` would be more appropriate: https://developer.mozilla.org/en-US/docs/Web/CSS/cursor.

besides this and a comment about modularity, the code looks great! i tested the functionality on the dev app and it works well :) awesome job

ysl-ee commented:
I worked on some minor details (spacing in code, and simplifying the sort_by variable update) and realized that the sort was not working on item requests (D:) and addressed that.

Currently still need to work on:
- Cursor should change to pointer on dropdown
- Using helper methods to simplify conditional block -> I did move things around to the helper methods and they do compile and run, but gets stuck. The most recent commit has the calls to the helper methods commented out, so the website should function.

@taylor-yamashita

Danielpark1239 commented:
There was an issue where the price sorting for item requests was going the wrong way, but after redeploying on the dev app, it seems to be working correctly.

Overall, this is a great feature and I hope it makes it easy to add further sorts down the line (ex. by quality or expiration date)! Looking forward to approving once the requested changes are made.

Danielpark1239 reviewed (changes requested):
same comments as Nick and Shannon, nothing new

taylor-yamashita commented:
> small nit: could you change cursor behavior when you hover a dropdown item? right now it's the default link cursor, but `cursor: pointer;` would be more appropriate: https://developer.mozilla.org/en-US/docs/Web/CSS/cursor.
> 
> besides this and a comment about modularity, the code looks great! i tested the functionality on the dev app and it works well :) awesome job

resolved! the other dropdown menus seem to inherit this property from css .btn, so added that for consistency. there is no specific css styling that targets dropdown menus; it all seems to be handled by bootstrap classes and directly in the html.

taylor-yamashita commented:
> I worked on some minor details (spacing in code, and simplifying the sort_by variable update) and realized that the sort was not working on item requests (D:) and addressed that.
> 
> Currently still need to work on:
> 
> * Cursor should change to pointer on dropdown
> * Using helper methods to simplify conditional block -> I did move things around to the helper methods and they do compile and run, but gets stuck. The most recent commit has the calls to the helper methods commented out, so the website should function.
> 
> @taylor-yamashita

will probably need to sync with @ysl-ee on the second one.

taylor-yamashita commented:
> @ysl-ee @taylor-yamashita could one of you upload a short screen recording or a series of screenshots to demonstrate the working functionality? for archival purposes :D

Added screenshots to the main PR comment! Let me know if you'd like it to be more detailed.

taylor-yamashita commented:
Condensed the "if" blocks in getItemsRelative() and getItemRequestsRelative() as per @nicholaspad 's suggestion. I looked at @ysl-ee 's helper methods (sortMyItems and ordering), moved the ordering() code back to the original methods (for now), and deleted duplicate code so the app doesn't break right now.

I know @ysl-ee was trying to make one helper function called by both getItemsRelative() and getItemRequestsRelative(), as per @shannon-heh 's suggestion, but honestly getItemsRelative() and getItemRequestsRelative() are themselves basically identical, so I think if we refactor this, we might want to consider combining those two whole functions into something more general instead of making a bunch of smaller helper functions. I haven't looked into it too much but it seems like the only difference is in variable names/constructing queries. 

If we do decide to refactor getItemsRelative() and getItemRequestsRelative(), I think it would be best for us to put that in a whole separate PR in the future, since there could be a lot of refactoring throughout the backend ("Items for Sale" and "Items Requested" pages mostly work the same, but I think there is a lot of separate duplicate code between them right now). Welcoming any thoughts!

nicholaspad reviewed (approved):
> If we do decide to refactor getItemsRelative() and getItemRequestsRelative(), I think it would be best for us to put that in a whole separate refactoring PR in the future, since there could be a lot of refactoring throughout the backend ("Items for Sale" and "Items Requested" pages mostly work the same, but I think there is a lot of separate duplicate code between them right now).

Completely agree with you about doing a larger refactoring in a future PR since `getItemsRelative` and `getItemRequestsRelative` are almost entirely the same. Thanks for adding this feature!! 🎉 

I'd wait for @shannon-heh or @Danielpark1239 to give one more approval before you merge this!

Danielpark1239 commented:
> Condensed the "if" blocks in getItemsRelative() and getItemRequestsRelative() as per @nicholaspad 's suggestion. I looked at @ysl-ee 's helper methods (sortMyItems and ordering), moved the ordering() code back to the original methods (for now), and deleted duplicate code so the app doesn't break right now.
> 
> I know @ysl-ee was trying to make one helper function called by both getItemsRelative() and getItemRequestsRelative(), as per @shannon-heh 's suggestion, but honestly getItemsRelative() and getItemRequestsRelative() are themselves basically identical, so I think if we refactor this, we might want to consider combining those two whole functions into something more general instead of making a bunch of smaller helper functions. I haven't looked into it too much but it seems like the only difference is in variable names/constructing queries.
> 
> If we do decide to refactor getItemsRelative() and getItemRequestsRelative(), I think it would be best for us to put that in a whole separate refactoring PR in the future, since there could be a lot of refactoring throughout the backend ("Items for Sale" and "Items Requested" pages mostly work the same, but I think there is a lot of separate duplicate code between them right now). Welcoming any thoughts!

I also agree that there are a lot of places where refactoring could help simplify the codebase, but that's not an urgent issue. Thanks for the changes! They look good to me, and I think the sort feature's ready to be merged. 👍

shannon-heh reviewed (approved):
awesome job!!! thanks for making the changes. i'll merge in the PR rn
